### ISSUE TITLE
(Some) callbacks are not being called during multi-GPU training

### ISSUE THREAD

**Issue by haimat** [ultralytics/ultralytics#8582]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Train, Multi-GPU

### Bug

It used to work fine in the past, but today I noticed that during multi-GPU training (some) callbacks are not being called anymore.

### Environment

Ultralytics YOLOv8.1.20 🚀 Python-3.10.12 torch-2.0.1+cu117 CUDA:0 (NVIDIA RTX A6000, 48677MiB)
Setup complete ✅ (48 CPUs, 251.6 GB RAM, 243.8/877.5 GB disk)

OS                  Linux-5.15.0-84-generic-x86_64-with-glibc2.35
Environment         Linux
Python              3.10.12
Install             pip
RAM                 251.56 GB
CPU                 AMD EPYC 7443 24-Core Processor
CUDA                11.7

matplotlib          ✅ 3.5.1>=3.3.0
opencv-python       ✅ 4.8.0.76>=4.6.0
pillow              ✅ 9.5.0>=7.1.2
pyyaml              ✅ 5.4.1>=5.3.1
requests            ✅ 2.25.1>=2.23.0
scipy               ✅ 1.8.0>=1.4.1
torch               ✅ 2.0.1>=1.8.0
torchvision         ✅ 0.15.2>=0.9.0
tqdm                ✅ 4.66.1>=4.64.0
psutil              ✅ 5.9.0
py-cpuinfo          ✅ 9.0.0
thop                ✅ 0.1.1-2209072238>=0.1.1
pandas              ✅ 2.1.1>=1.1.4
seaborn             ✅ 0.12.2>=0.11.0

### Minimal Reproducible Example

```python
from ultralytics import YOLO

def yolo_cb_batch_end(trainer):
    print("\n>>>yolo_cb_batch_end")  # this output is only printed when trained on one GPU

def yolo_cb_train_end(trainer):
    print("\n>>>yolo_cb_train_end")  # this output is only printed when trained on one GPU

if __name__ == "__main__":
    model = YOLO("yolov8n")
    model.add_callback("on_batch_end", yolo_cb_batch_end)
    model.add_callback("on_train_end", yolo_cb_train_end)

    # Start model training with the parameters from the cvServer UI
    model.train(
        data="coco128.yaml",
        device="0,1,2,3",  # if I put "0" here then I can see the output from the callbacks
        batch=32,
        imgsz=640,
        epochs=3,
        project="/tmp",
        name="",
        exist_ok=True,
        verbose=False,
        val=False,
    )
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by haimat** [comment]
A solution is discussed in https://github.com/ultralytics/ultralytics/issues/6168

**Comment by glenn-jocher** [comment]
@haimat thank you for bringing this to our attention! It seems like you're encountering an issue with callbacks not being called during multi-GPU training. This is indeed a peculiar behavior that we need to look into. 

As you've noticed, callbacks are expected to work across both single and multi-GPU setups. The behavior you're experiencing might be related to how PyTorch's Distributed Data Parallel (DDP) handles processes across multiple GPUs. Each process in a multi-GPU setup operates independently, which can sometimes lead to unexpected behavior with callbacks.

A similar issue was discussed and addressed in [#6168](https://github.com/ultralytics/ultralytics/issues/6168). The solution involved ensuring that callbacks are correctly managed across all processes in a DDP setup. Here's a quick snippet on how you might adjust your callback functions to ensure they're being called correctly:

```python
def yolo_cb_batch_end(trainer):
    if trainer.is_main_process:
        print("\n>>>yolo_cb_batch_end")

def yolo_cb_train_end(trainer):
    if trainer.is_main_process:
        print("\n>>>yolo_cb_train_end")
```

By checking `trainer.is_main_process`, you ensure that the callback is executed only by the main process in a multi-GPU training session. This should help in managing the callback executions properly across all GPUs.

Please give this a try and let us know if it resolves the issue. If the problem persists, feel free to share more details, and we'll be happy to dive deeper into this. Your contributions and feedback are invaluable to the YOLO community and the Ultralytics team! 🚀

**Comment by glenn-jocher** [comment]
@rromanchuk hi there! 😊 It looks like there was a bit of a mix-up. The `is_main_process` attribute I mentioned doesn't directly exist in the Ultralytics YOLOv8 repository as of now. My apologies for the confusion. 

For multi-GPU setups, a similar effect can often be achieved by checking the rank of the process in distributed training, typically using something like `torch.distributed.get_rank()` to ensure certain operations only happen in the main process. 

For tailored advice fitting the current Ultralytics YOLOv8 framework, I recommend checking the latest docs or the source for handling callbacks in a distributed setting. Thanks for catching that, and please let me know if there's more I can help with!
